File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
    const res: number[][] = [], q: [TreeNode, number][] = [[root, 0]]
    if (!root) return res
    traverse()
    return res
    
    function traverse() {
        if (!q.length) return
        const [{val, left, right}, lvl] = q.shift()
        res[lvl] = res[lvl] || []
        res[lvl].push(val)
        if (left) q.push([left, lvl+1])
        if (right) q.push([right, lvl+1])
        traverse()
    }
};

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val res = mutableListOf<List<Int>>()
        if (root === null) {
            return res
        }
        val q: Queue<TreeNode> = LinkedList()
        q.offer(root)
        while (q.isNotEmpty()) {
            val len = q.size
            val lvl = mutableListOf<Int>()
            for (i in 0..len-1) {
                val cur = q.poll()
                lvl.add(cur.`val`)
                if (cur.`left` !== null) {
                    q.offer(cur.`left`)
                }
                if (cur.`right` !== null) {
                    q.offer(cur.`right`)
                }
            }
            res.add(lvl)
        }
        return res
    }
}